Update PandasUDF doc (#2089)
NVIDIA/spark-rapids: docs/additional-functionality/rapids-udfs.md

[native code](../../udf-examples/src/main/cpp/src) to count words in strings

> ## GPU support for Pandas UDF

Reviewer: ```suggestion
## GPU Support for Pandas UDF
```

Author: Updated.

---
**NOTE**

> The _GPU support for Pandas UDF_ is an experimental feature, and may change at any point it time.

Reviewer: Italics are out of place here, "GPU support for Pandas UDF" is not an official name for anything.
```suggestion
The GPU support for Pandas UDF is an experimental feature, and may change at any point it time.
```

Author: Updated.

---

> _GPU support for Pandas UDF_ is built on Apache Spark's [Pandas UDF(user defined

Reviewer: ```suggestion
GPU support for Pandas UDF is built on Apache Spark's [Pandas UDF(user defined
```

Author: Updated.

function)](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#pandas-udfs-a-k-a-vectorized-udfs),
and has two features:

> - **GPU Assignment(Scheduling) in Python Process**: Let the Python process share the same GPU with Spark executor JVM. Without this feature, some use case in PandasUDF(a.k.a an `independent` process) will likely to use other GPUs other than the one we want it to run on. e.g. user can launch a TensorFlow session inside Pandas UDF and the machine contains 8 GPUs. user launchs 8 Spark executors. Without this GPU sharing feature, TensorFlow will automatically use all 8 GPUs it can detects which will definitly conflict with existing Spark executor JVM processes.

Reviewer 2: ```suggestion
- **GPU Assignment(Scheduling) in Python Process**: Let the Python process share the same GPU with Spark executor JVM. Without this feature, in a non-isolated environment, some use cases with PandasUDF (a.k.a an `independent` process) can try to use GPUs other than the one we want it to run on. For example, the user could launch a TensorFlow session inside Pandas UDF and the machine contains 8 GPUs. Without this GPU sharing feature, TensorFlow will automatically use all 8 GPUs which will conflict with existing Spark executor JVM processes.
```

Author: Done.

> - **GPU Assignment(Scheduling) in Python Process**: Let the Python process share the same GPU with Spark executor JVM. Without this feature, some use case in PandasUDF(a.k.a an `independent` process) will likely to use other GPUs other than the one we want it to run on. e.g. user can launch a TensorFlow session inside Pandas UDF and the machine contains 8 GPUs. user launchs 8 Spark executors. Without this GPU sharing feature, TensorFlow will automatically use all 8 GPUs it can detects which will definitly conflict with existing Spark executor JVM processes.

Reviewer: Please wrap lines to approx. 100 characters to increase readability when examining the documentation in a terminal and to make editing/commenting easier.  Edits to this really long line below:
```suggestion
- **GPU Assignment(Scheduling) in Python Process**: Lets the Python process share the same GPU
with the Spark executor JVM process. Without this feature, GPU usage in a Pandas UDF could use
a different GPU than the one used by the Spark executor.  As an example, consider a TensorFlow
session launched inside a Pandas UDF and the machine contains 8 GPUs. The user launches 8 Spark
executors. Without this GPU sharing feature, TensorFlow will automatically use all 8 GPUs it detects
which will conflict with the Spark executor JVM processes.
```

Author: Thanks for this point! Updated.

> - **Increase Speed**: Speed up data transfer between JVM process and Python process.

Reviewer: ```suggestion
- **Increase Speed**: Speeds up data transfer between JVM process and Python process.
```

Author: Updated.

> To enable _GPU support for Pandas UDF_, you need to configure your spark job with extra settings.

Reviewer: ```suggestion
To enable GPU support for Pandas UDF, you need to configure your Spark job with extra settings.
```

1. Make sure GPU `exclusive` mode is disabled. Note that this will not work if you are using exclusive

>    mode to assign GPUs under spark.

Reviewer: ```suggestion
   mode to assign GPUs under Spark.
```

> 2. Currently the python files are packed into the spark rapids plugin jar.

Reviewer: ```suggestion
2. Currently the Python files are packed into the RAPIDS Accelerator jar.
```

[native code](../../udf-examples/src/main/cpp/src) to count words in strings

## GPU Support for Pandas UDF

---
**NOTE**

> The GPU Support for Pandas UDF is an experimental feature, and may change at any point it time.

Author: Done

---

GPU support for Pandas UDF is built on Apache Spark's [Pandas UDF(user defined
function)](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#pandas-udfs-a-k-a-vectorized-udfs),
and has two features:

- **GPU Assignment(Scheduling) in Python Process**: Let the Python process share the same GPU with
Spark executor JVM. Without this feature, in a non-isolated environment, some use cases with
PandasUDF (a.k.a an `independent` process) can try to use GPUs other than the one we want it to
run on. For example, the user could launch a TensorFlow session inside Pandas UDF and the machine
contains 8 GPUs. Without this GPU sharing feature, TensorFlow will automatically use all 8 GPUs
which will conflict with existing Spark executor JVM processes.

- **Increase Speed**: Speeds up data transfer between JVM process and Python process.

> To enable GPU support for Pandas UDF, you need to configure your spark job with extra settings.

Author: Updated.

> 1. Make sure GPU `exclusive` mode is disabled. Note that this will not work if you are using exclusive

Reviewer 3: For On-prem YARN instructions, we add a line "refer to nvidia-smi documentation". 

Do we need to add the link or sample like this? 
nvidia-smi -i 0 -c EXCLUSIVE_PROCESS # Set GPU 0 to exclusive mode, run as root.

Also, if this mandatory for YARN, should we say that more explicitly?

Author: Thanks for review!
Added a line to show the command to set GPU to Default mode.
I think no need to say if it's for yarn or for other environment. The _mode_ here is only for GPU hardware. If the GPU is in Default mode, it will always be default mode no matter on Yarn or on Standalone or Local.

    --py-files ${SPARK_RAPIDS_PLUGIN_JAR}
    ```

3. Enable GPU Assignment(Scheduling) for Pandas UDF.

    ```shell
    ...
    --conf spark.rapids.python.gpu.enabled=true \
    ```

> Please note the data transfer acceleration only supports `scalar UDF` and `Scalar iterator UDF` currently. 

Reviewer: ```suggestion
Please note the data transfer acceleration only supports `Scalar UDF` and `Scalar iterator UDF` currently. 
```

> Please note the data transfer acceleration only supports `scalar UDF` and `Scalar iterator UDF` currently. 

Reviewer 4: We could have a table to show the status of the data transfer acceleration for all the 6 plans, instead of this notice. You can find the latest status from the commments of issue https://github.com/NVIDIA/spark-rapids/issues/305 .

Author: Good suggestion, made a status table for them.

> You could choose the exec see the section below.

Reviewer: What section below?  Either link to it explicitly or if it's the immediate section below then just remove this sentence.

### Other Configuration

> Following configuration settings are also for _GPU Scheduling for Pandas UDF_

Reviewer: ```suggestion
Following configuration settings are also relevant for GPU Scheduling for Pandas UDF.
```

1. Spark Plan support

    ```
    --conf spark.rapids.sql.exec.ArrowEvalPythonExec=true \

>     --conf spark.rapids.sql.exec.MapInPandasExec=false \

Reviewer 2: did you mean to have these false? or are these false due to performance?

Author: Removed and add a status table for them for better view.

    --conf spark.rapids.sql.exec.FlatMapGroupsInPandasExec=false \
    --conf spark.rapids.sql.exec.AggregateInPandasExec=false \
    --conf spark.rapids.sql.exec.FlatMapCoGroupsInPandasExec=false \

>     --conf spark.rapids.sql.exec.WindowInPandasExec=true

Reviewer: Most of these settings match the defaults, so do we need to burden the user with these details?  There's no explanation in this documentation why these settings are necessary.

Author: You got reason. I made a table to show the their status at data-transfer part.

    ```

>     These configs are the switches for each type of PandasUDF execution plan. Some of theme are set to false by default due to not supported or performance issue.

Reviewer 2: s/theme/them/

Author: Removed as suggested.

>     These configs are the switches for each type of PandasUDF execution plan. Some of theme are set to false by default due to not supported or performance issue.

Reviewer: Should be wrapped to approx. 100 characters.

Author: Updated.

2. Memory efficiency

    ```
    --conf spark.rapids.python.memory.gpu.pooling.enabled=false \
    --conf spark.rapids.python.memory.gpu.allocFraction=0.1 \
    --conf spark.rapids.python.memory.gpu.maxAllocFraction= 0.2 \
    ```

>     Same to the [RMM pooling for JVM](../tuning-guide.md#pooled-memory), here the pooling serves the same way but for Python process.  `half of the rest GPU memory` will be used by default if it is not specified.

Reviewer 2: to be clear here, is this the python process will assume it can use half the memory of the GPU?  I think we should mention the other spark.rapids.memory.gpu.allocFraction setting here.

Author: Updated.

>     Same to the [RMM pooling for JVM](../tuning-guide.md#pooled-memory), here the pooling serves the same way but for Python process.  `half of the rest GPU memory` will be used by default if it is not specified.

Reviewer: Wrap to 100 and make it the first paragraph to introduce the config settings.
```suggestion
Similar to the [RMM pooling for JVM](../tuning-guide.md#pooled-memory) settings except these specify the
GPU pool size for the Python process.  Half of the GPU available memory will be used by default if it is not specified.
```

Author: Updated.

3. Limit of concurrent Python processes

    ```
    --conf spark.rapids.python.concurrentPythonWorkers=2 \
    ```

>     This parameter aims to limit the total concurrent running `Python process` in 1 Spark executor. This parameter is set to 0 by default which means there's not limit for concurrent Python workers. Note that for certain cases, setting this value too small may result a `hang` for your Spark job because a PandasUDF may produces multiple python process and each will try to acquire the python GPU process semaphore. This may bring a dead lock situation becasue a Spark job will not preceed until all its tasks are finished. For example, in the Pandas UDF stage, 2 tasks are running and each task launches 3 Python process and we set this parameter to 4.

Reviewer 2: I'm a little confused by this, how do I know how many python processes I need?  On spark I usually get one per python process per task. Can we add more detail here

Author: I made wrong description here, sorry for that. Discussed with Liangcai today and re-state with more specific example.

>     This parameter aims to limit the total concurrent running `Python process` in 1 Spark executor. This parameter is set to 0 by default which means there's not limit for concurrent Python workers. Note that for certain cases, setting this value too small may result a `hang` for your Spark job because a PandasUDF may produces multiple python process and each will try to acquire the python GPU process semaphore. This may bring a dead lock situation becasue a Spark job will not preceed until all its tasks are finished. For example, in the Pandas UDF stage, 2 tasks are running and each task launches 3 Python process and we set this parameter to 4.

Reviewer: Wrap to 100 chars.
```suggestion
This parameter limits the total concurrent running `Python process` associated with a Spark executor.
The parameter defaults to 0 which means no limit. Note that for certain cases, setting this value too small
may result a hang for your Spark job.  A PandasUDF may produce multiple Python process and
each will try to acquire the Python GPU process semaphore. This may result in a deadlock situation
because a Spark job will not proceed until all the tasks are finished. For example, in the Pandas UDF stage
2 tasks are running, and each task launches 3 Python process and we set this parameter to 4.
```

Author: Updated.

    ![Python concurrent worker](/docs/img/concurrentPythonWorker.PNG)

>     Two python process in each task acquired their semaphore but neither of them are able to proceed becasue both of them are waiting for their third semaphore to acutally start the task. 

Reviewer: ```suggestion
Two Python process in each task acquired their semaphore, but neither of them are able
to proceed because both of them are waiting for their third semaphore to start the task. 
```

Author: Done.

    On Standalone, you need to add
    ```shell
    ...
    --conf spark.executorEnv.PYTHONPATH=${SPARK_RAPIDS_PLUGIN_JAR} \
    --py-files ${SPARK_RAPIDS_PLUGIN_JAR}
    ```

3. Enable GPU Assignment(Scheduling) for Pandas UDF.

    ```shell
    ...
    --conf spark.rapids.python.gpu.enabled=true \
    ```

Please note: every type of Pandas UDF on Spark is run by a specific Spark execution plan. RAPIDS
Accelerator has a 1-1 mapping support for each of them. Not all Pandas UDF types are data-transfer
accelerated at present:

  | Spark Execution Plan|Data Transfer Accelerated|Use Case|
  |----------------------|----------|--------|

>   |ArrowEvalPythonExec|yes|[Series to Series](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#series-to-series), [Iterator of Series to Iterator of Series](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#iterator-of-series-to-iterator-of-series) and [Iterator of Multiple Series to Iterator of Series](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#iterator-of-multiple-series-to-iterator-of-series)| supported|

Reviewer 5: What does the `| supported|` mean in this case?  It is a fourth column in a three column table.

Author: wow, didn't notice that it doesn't show in the preview. Thanks for catching it. Updated.

  |MapInPandasExec|yes|[Map](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#map)|
  | WindowInPandasExec|yes|[Window](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#series-to-scalar)|
  | FlatMapGroupsInPandasExec|no|[Grouped Map](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#grouped-map)|
  | AggregateInPandasExec|no|[Aggregate](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#series-to-scalar)|
  |FlatMapCoGroupsInPandasExec|no|[Co-grouped Map](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#co-grouped-map)|

### Other Configuration

The following configuration settings are also relevant for GPU scheduling for Pandas UDF.

1. Memory efficiency

>     ```

Reviewer 5: ```suggestion
    ```shell
```

    --conf spark.rapids.python.memory.gpu.pooling.enabled=false \
    --conf spark.rapids.python.memory.gpu.allocFraction=0.1 \
    --conf spark.rapids.python.memory.gpu.maxAllocFraction= 0.2 \
    ```
    Similar to the [RMM pooling for JVM](../tuning-guide.md#pooled-memory) settings like
    `spark.rapids.memory.gpu.allocFraction` and `spark.rapids.memory.gpu.maxAllocFraction` except
    these specify the GPU pool size for the _Python process_. Half of the GPU _available_ memory
    will be used by default if it is not specified.

2. Limit of concurrent Python processes

>     ```

Reviewer 5: ```suggestion
    ```shell
```

Author: Updated.

    deadlock situation because a Spark job will not proceed until all its tasks are finished.

    For example, in a specific Spark Stage that contais 3 PandasUDFs, 2 Spark tasks are running and
    each task launches 3 Python process while we set this `concurrentPythonWorkers` to 4.

    ```python
    df_1 = df_0.mapInPandas(udf_1, schema_1)
    df_2 = df_1.mapInPandas(udf_2, schema_2)
    df_3 = df_2.mapInPandas(udf_3, schema_3)
    df_3.explain(True)
    ```
    The RAPIDS Accelerator query explain:
    ```
    ...
      *Exec <MapInPandasExec> could partially run on GPU
        *Exec <MapInPandasExec> could partially run on GPU
          *Exec <MapInPandasExec> could partially run on GPU
    ...
    ```

    ![Python concurrent worker](../img/concurrentPythonWorker.PNG)

    In this case, each PandasUDF will launch a Python process. At this moment two Python process
    in each task(in light green) acquired their semaphore but neither of them are able to proceed
    because both of them are waiting for their third semaphore to start the task.

    Another example is to use ArrowEvalPythonExec, with the following code:

>     ```

Reviewer 5: ```suggestion
    ```python
```

    import pyspark.sql.functions as F
    ...
    df = df.withColumn("c_1",udf_1(F.col("a"), F.col("b")))
    df = df.withColumn('c_2', F.hash(F.col('c_1')))
    df = df.withColumn("c_3",udf_2(F.col("c_2")))
    ...
    ```
    The physical plan:
    ```
        +- GpuArrowEvalPython
          +- ...
            +- ...
              +- GpuArrowEvalPython
    ```
    This means each Spark task will trigger 2 Python processes. In this case, if we set
    `concurrentPythonWorkers=2`, it will also probably result in a hang as we allow 2 tasks running

>     and each of them has 2 Python processes. Let's say Task_1_Process_1 and Task_2_Process_1

Author: Thanks for review! Updated.

    --py-files ${SPARK_RAPIDS_PLUGIN_JAR}
    ```

3. Enable GPU Assignment(Scheduling) for Pandas UDF.

    ```shell
    ...
    --conf spark.rapids.python.gpu.enabled=true \
    ```

Please note: every type of PandasUDF on Spark is run by a specific Spark execution plan. RAPIDS
Accelerator has a 1-1 mapping support for each of them. Not all PandasUDF types are data-transfer
accelerated at present:

>   | Spark Execution Plan|Data Transfer accelerated|Use Case|

Reviewer 4: nit:
```suggestion
  | Spark Execution Plan|Data Transfer Accelerated|Use Case|
```

1. Memory efficiency

    ```
    --conf spark.rapids.python.memory.gpu.pooling.enabled=false \
    --conf spark.rapids.python.memory.gpu.allocFraction=0.1 \
    --conf spark.rapids.python.memory.gpu.maxAllocFraction= 0.2 \
    ```
    Similar to the [RMM pooling for JVM](../tuning-guide.md#pooled-memory) settings like
    `spark.rapids.memory.gpu.allocFraction` and `spark.rapids.memory.gpu.maxAllocFraction` except
    these specify the GPU pool size for the _Python process_. Half of the GPU _available_ memory
    will be used by default if it is not specified.

2. Limit of concurrent Python processes

    ```
    --conf spark.rapids.python.concurrentPythonWorkers=2 \
    ```
    This parameter limits the total concurrent running `Python process` for a Spark executor.
    It defaults to 0 which means no limit. Note that for certain cases, setting
    this value too small _may result in  a hang for your Spark job_ because a task may contain
    multiple PandasUDF(`MapInPandas`) instances which results in multiple Python processes.
    Each process will try to acquire the Python GPU process semaphore. This may result in a

>     deadlock situation because a Spark job will not proceed until all its tasks are finished

Reviewer: ```suggestion
    deadlock situation because a Spark job will not proceed until all its tasks are finished.
```

    For example, in a specific Spark Stage that contais 3 PandasUDFs, 2 Spark tasks are running and
    each task launches 3 Python process while we set this `concurrentPythonWorkers` to 4.

    ```python
    df_1 = df_0.mapInPandas(udf_1, schema_1)
    df_2 = df_1.mapInPandas(udf_2, schema_2)
    df_3 = df_2.mapInPandas(udf_3, schema_3)
    df_3.explain(True)
    ```
    The RAPIDS Accelerator query explain:
    ```
    ...
      *Exec <MapInPandasExec> could partially run on GPU
        *Exec <MapInPandasExec> could partially run on GPU
          *Exec <MapInPandasExec> could partially run on GPU
    ...
    ```

>     ![Python concurrent worker](/docs/img/concurrentPythonWorker.PNG)

Reviewer 5: The image does not show up on the jekyll rendered page.  Try changing to: 
```suggestion
    ![Python concurrent worker](../img/concurrentPythonWorker.PNG)
```

Author: Done.

    --py-files ${SPARK_RAPIDS_PLUGIN_JAR}
    ```

3. Enable GPU Assignment(Scheduling) for Pandas UDF.

    ```shell
    ...
    --conf spark.rapids.python.gpu.enabled=true \
    ```

Please note: every type of PandasUDF on Spark is run by a specific Spark execution plan. RAPIDS
Accelerator has a 1-1 mapping support for each of them. Not all PandasUDF types are data-transfer
accelerated at present:

>   | Spark Execution Plan | Use Case | Status |

Reviewer 2: can we change Status to be data transfer accelerated or accelerated just to be more clear

Author: Use data transfer accelerated.

  |----------------------|----------|--------|
  |ArrowEvalPythonExec|[Series to Series](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#series-to-series), [Iterator of Series to Iterator of Series](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#series-to-series) and [Iterator of Multiple Series to Iterator of Series](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#iterator-of-multiple-series-to-iterator-of-series)| supported|
  |MapInPandasExec| [Map](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#map)| supported|

>   | WindowInPandasExec | [Window](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#series-to-scalar)| supported|

Reviewer 2: nit looks like we have extra spaces in a couple of these, we may also switch that column to be second and have use case third

Author: switched.

  | FlatMapGroupsInPandasExec| [Grouped Map](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#grouped-map)| not supported|
  | AggregateInPandasExec| [Aggregate](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#series-to-scalar)| not supported|
  |FlatMapCoGroupsInPandasExec| [Co-grouped Map](https://spark.apache.org/docs/latest/api/python/user_guide/arrow_pandas.html#co-grouped-map)| not supported|

### Other Configuration

> Following configuration settings are also relevant for GPU Scheduling for Pandas UDF.

Author: Updated.

1. Memory efficiency

    ```
    --conf spark.rapids.python.memory.gpu.pooling.enabled=false \
    --conf spark.rapids.python.memory.gpu.allocFraction=0.1 \
    --conf spark.rapids.python.memory.gpu.maxAllocFraction= 0.2 \
    ```
    Similar to the [RMM pooling for JVM](../tuning-guide.md#pooled-memory) settings like
    `spark.rapids.memory.gpu.allocFraction` and `spark.rapids.memory.gpu.maxAllocFraction` except

>     these specify the GPU pool size for the `Python process`. Half of the GPU `available` memory

Author: Noted. Updated.

    will be used by default if it is not specified.

2. Limit of concurrent Python processes

    ```
    --conf spark.rapids.python.concurrentPythonWorkers=2 \
    ```
    This parameter aims to limit the total concurrent running `Python process` in 1 Spark executor.
    This parameter is set to 0 by default which means there's not limit for concurrent Python
    workers. Note that for certain cases, setting this value too small may result a `hang` for
    your Spark job because a Spark Task may contain multiple PandasUDF(`MapInPandas`) which results
    multiple python processes and each will try to acquire the python GPU process semaphore. This

>     may bring a dead lock situation becasue a Spark job will not preceed until all its tasks are

Author: Refined.

    finished.

    For example, in a specific Spark Stage that contais 3 PandasUDFs, 2 Spark tasks are running and
    each task launches 3 Python process while we set this `concurrentPythonWorkers` to 4.

    ```python
    df_1 = df_0.mapInPandas(udf_1, schema_1)
    df_2 = df_1.mapInPandas(udf_2, schema_2)
    df_3 = df_2.mapInPandas(udf_3, schema_3)
    df_3.explain(True)
    ```
    The RAPIDS Accelerator query explain:
    ```
    ...
      *Exec <MapInPandasExec> could partially run on GPU
        *Exec <MapInPandasExec> could partially run on GPU
          *Exec <MapInPandasExec> could partially run on GPU
    ...
    ```

    ![Python concurrent worker](/docs/img/concurrentPythonWorker.PNG)

>     In this case, each PandasUDF will launch a Python process. At this moment two python process

Author: Updated.

>     in each task acquired their semaphore but neither of them are able to proceed becasue both

Reviewer 2: because is misspelled

Author: Done.
